src/test/java/codeu/controller/AdminServletTest.java

```diff
@@ -107,4 +107,58 @@ public void testDoGet_UserIsAdmin() throws IOException, ServletException {
         Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
     }
 
+    @Test
+    public void testDoPost() throws IOException, ServletException {
+        adminServlet.doPost(mockRequest, mockResponse);
+        Assert.assertNull(mockRequest.getParameter("deleteUsersButton"));
+        Assert.assertNull(mockRequest.getParameter("deleteMessagesButton"));
+        Assert.assertNull(mockRequest.getParameter("deleteConversationsButton"));
+
+        Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+    }
+
+    @Test
+    public void testDoPost_DeleteUsers() throws ServletException, IOException {
+        Assert.assertEquals(1234, mockMessageStore.getNumMessages());
+        Assert.assertEquals(12345, mockUserStore.getNumUsers());
+        Mockito.when(mockRequest.getParameter("deleteUsersButton")).thenReturn("notNull");
+
+        adminServlet.doPost(mockRequest, mockResponse);
+
+        mockMessageStore.deleteAllMessages();
```

Comment:
Shouldn't you be doing Mockito.verify(mockMessageStore).deleteAllMessages();

Similarly in other places

Reply:
As my code is now, I can't do Mockito.verify on the stores because they're test instances and not mocks of the store classes. I wasn't sure how to quite handle this even when I finished my MVP; I couldn't mock the actual stores themselves and verify the method calls because they were singletons accessed through getInstance()
For example, if I make my test this:
![image](https://user-images.githubusercontent.com/11599574/40879325-6ec865b2-665b-11e8-9908-2a8f2ad568bf.png)

I get the error "Wanted but not invoked: messageStore.deleteAllMessages();". 
So that the reason why I switched to using the Store test instances provided by the class which uses a mock PersistentStorageAgent to construct (which I'm not sure how to verify method calls with).

Reply:
Ah, understood. I got a bit confused by the naming. I think it's probably better to rename it `fakeMessageStore`. This [article](https://martinfowler.com/articles/mocksArentStubs.html) gives a pretty good explanation on the different kinds of test doubles. 

Also, if I understand correctly and you're not verifying behaviour, you shouldn't be calling these statements since that's what the `AdminServlet` is supposed to do.

Anyways, it's a good thing that you are verifying the state with assertions 👍

Reply:
Just want to double check; when you say I "shouldn't be calling these statements" which ones are you referring to?

Reply:
Ah okay, I see what you mean now.
I'm sort of confused on how to actually get it so that I can verify the actual stores running their delete methods though, since they're singletons I can't mock the class itself to verify method calls because all of the real calls go through getInstance() first and something like Mockito.when(MessageStore.getInstance()).thenReturn(Mockito.mock(MessageStore.class)); wouldn't work since I can't mock the static method. Do you have any suggestions on how I should approach this?

Reply:
So, with testing there's two broad strategies you can take,
1. Verify behaviour by using mocks and verifying that certain methods are called
2. Verify state by using fakes and asserting that the end state is what is desired

Although both approaches have their applications, verifying state is generally considered the better approach since verifying behaviour is brittle and a small change in which methods you call can cause your test to fail, even though you only really care about the end-result.

It looks like you are already verifying state with the statements,
```
Assert.assertEquals(0, mockMessageStore.getNumMessages());
Assert.assertEquals(0, mockUserStore.getNumUsers());
```

So, I'd say that you're good.